src/WebJobs.Extensions.DurableTask/ContextInterfaces/IDurableEntityClient.cs

```diff
@@ -43,6 +43,44 @@ public interface IDurableEntityClient
         /// <returns>A task that completes when the message has been reliably enqueued.</returns>
         Task SignalEntityAsync(EntityId entityId, DateTime scheduledTimeUtc, string operationName, object operationInput = null, string taskHubName = null, string connectionName = null);
 
+        /// <summary>
+        /// Signals an entity to perform an operation.
+        /// </summary>
+        /// <typeparam name="TEntityInterface">Entity interface.</typeparam>
+        /// <param name="entityKey">The target entity key.</param>
+        /// <param name="operation">A delegate that performs the desired operation on the entity.</param>
+        /// <returns>A task that completes when the message has been reliably enqueued.</returns>
+        Task SignalEntityAsync<TEntityInterface>(string entityKey, Action<TEntityInterface> operation);
+
+        /// <summary>
+        /// Signals an entity to perform an operation, at a specified time.
+        /// </summary>
+        /// <typeparam name="TEntityInterface">Entity interface.</typeparam>
+        /// <param name="entityKey">The target entity key.</param>
+        /// <param name="scheduledTimeUtc">The time at which to start the operation.</param>
+        /// <param name="operation">A delegate that performs the desired operation on the entity.</param>
+        /// <returns>A task that completes when the message has been reliably enqueued.</returns>
+        Task SignalEntityAsync<TEntityInterface>(string entityKey, DateTime scheduledTimeUtc, Action<TEntityInterface> operation);
+
+        /// <summary>
+        /// Signals an entity to perform an operation.
+        /// </summary>
+        /// <typeparam name="TEntityInterface">Entity interface.</typeparam>
+        /// <param name="entityId">The target entity.</param>
+        /// <param name="operation">A delegate that performs the desired operation on the entity.</param>
+        /// <returns>A task that completes when the message has been reliably enqueued.</returns>
+        Task SignalEntityAsync<TEntityInterface>(EntityId entityId, Action<TEntityInterface> operation);
+
+        /// <summary>
+        /// Signals an entity to perform an operation, at a specified time.
+        /// </summary>
+        /// <typeparam name="TEntityInterface">Entity interface.</typeparam>
+        /// <param name="entityId">The target entity.</param>
+        /// <param name="scheduledTimeUtc">The time at which to start the operation.</param>
+        /// <param name="operation">A delegate that performs the desired operation on the entity.</param>
+        /// <returns>A task that completes when the message has been reliably enqueued.</returns>
+        Task SignalEntityAsync<TEntityInterface>(EntityId entityId, DateTime scheduledTimeUtc, Action<TEntityInterface> operation);
```

Comment:
Have you considered using optional parameters and creating a smaller number of method overloads for things like this? I feel like it would be much easier for a developer to mock a single method rather than four. It also reduces our documentation burden. Is this an option we can safely consider given we're already planning on making a breaking change to these interfaces?

This question applies to all the interfaces you're updating (entity client, orchestration client, orchestration context, etc.)

Reply:
So I considered handling default parameters on this interfaces, but I remember running into some oddness around default parameters with interfaces in the past. [This article](https://michael-mckenna.com/optional-arguments-in-c-sharp-interfaces/) explains why they can be dangerous. Though I did write all of these as explicit interface implementations, so it may not be an issue...

Even if it is an issue, since we only have one implementation, I think we could get away with it as long as we wrote a test that would catch these kind of errors using reflection.

Reply:
So initial investigations makes it look like this should not be a problem for our production code at least.

That being said, [another issue](https://stackoverflow.com/questions/12413583/an-expression-tree-may-not-contain-a-call-or-invocation-that-uses-optional-argum) popped up when I tested out using default parameters with Moq. Essentially, since optional parameters are compile time attributes, Moq can't handle them. The workaround is for customers to use `It.IsAny<T>()` in their `.Setup()` expression trees for all of the optional parameters, even if they don't use them in their production code. 

It is a bit more annoying as a test writer, but it is still a vast improvement to the previous problem, as intellisense gives you fairly clear instructions on how to fix the `.Setup()`.

So essentially the tradeoff is between:

1. Explicitly have all overloads as interface methods, and maintain the documentation + explicit interface implementations for them, in order to maximize test writeability.
2. Simplify our burden by using parameters, and make test writers explicitly handle the optional parameters in their Setup operations with help from intellisense.

I don't have strong opinions about which set of trade offs is better in this case.

Reply:
If you're open to either option, I would cast my vote for (2) to help us reduce our maintenance burdens.

Reply:
Also worth noting is that this makes adding a new optional parameter break customer tests, where has having them as an their own interface method with an implementation means customers using Moq won't be broken by adding that new method.

Though, it's fair to argue that adding optional parameters to these interfaces causes bloat to our API surface that would better be handled by some data object that contains all of the optional parameters, and that we should be hesitant about adding them regardless...

Reply:
*sigh*

If I understand correctly, you're saying this will basically break _everyone's_ test code since their existing Moq code would be incompatible? In that case, I think we have to go with your current approach.

But honestly, I'm already convinced we need to just move back to abstract classes to avoid all these problems with interfaces. I think we can do this without introducing any breaking changes - just have them implement the interfaces but have all future updates only target the abstract classes (and update all documentation to use the abstract classes instead of interfaces). That's how I would prefer we evolve our APIs moving forward. This interface stuff is becoming too difficult to deal with IMO.

Reply:
I don't think it would break most people's test code because the current interface method that customer's can mock is already the version with all the optional parameters provided, which is what they would have to mock with the default parameter approach. I would have to verify that is the case for most/all methods we would use default parameters.

Essentially what we are doing by making some of the parameters optional on the interface is we are giving them intellisense guidance when writing their Moq setup expressions; right now, they get runtime failures, or if they use a Moq analyzer, get compile time errors but they don't know which of the method overloads is the correct one to Moq.

I am not opposed to going back to abstract class implementations moving forward, but I think that abstract classes would have similar issues with Moq and optional parameters moving forward.

Essentially if we did something like adding an optional cancellation token to an API (an example I know we have done in the past), it would break Moq consumers even if it was a virtual method on an abstract class.

Reply:
Looking at this a second time, I'm realizing that my original interpretation of this change and the current state of the world was completely backwards! As you mentioned, the interface _already_ defined the most generic overload and used optional parameters. Therefore this is non-breaking for existing Moq-based test code. I think I now understand that the point of this PR is to add all the extension methods as explicit interface methods so the users can actually mock the overloads that they're using.